### PR TITLE
Allow v[k] on LHS of compound assignment statement

### LINKED ISSUE #415: Allow `v[k]` on LHS of compound assignment statement
Now that we support typed arrays/maps #328, it makes sense to allow LHR of a compound assignment statement to have the form `v[k]`.

### PR DESCRIPTION
**Pull request by ushirask**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

**Review by jclark — CHANGES_REQUESTED** [review]
Tests should be in 07-array.

You need tests for:

- map on LHS (e test in 07-map)
- array index is an expression
- evaluation order between array index and RHS (array index should be first)
- test to show that array index expression is being evaluated only once

**Review by jclark — CHANGES_REQUESTED** [review]
In

```
v[E1] += E2;
```

E1 needs to be evaluated once not twice. So you cannot simply transform it into

```
v[E1] = v[E1] + E2;
```

So you need to

1. Evaluate E1 into a register r1
2. Evaluate E2 into a register r2
3. ListGet(v, r1) into a register r3
5. Evaluate r3 + r2 into a register r4
5. ListSet(v, r1, r4)

File: compiler/modules/front/codeGen.bal [hunk 1]
```diff
@@ -858,6 +841,105 @@ function codeGenAssignToMember(CodeGenContext cx, bir:BasicBlock startBlock, Env
...
+        expr = { arithmeticOp: op, left: lValue, right: rexpr, pos: pos };
+    }
+    else {
+        expr = { bitwiseOp: <s:BinaryBitwiseOp> op, left: lValue, right: rexpr };
+    }
+    return codeGenAssignToVar(cx, startBlock, env, lValue.varName, expr);
+}
+
+function codeGenCompoundAssignToMember(CodeGenContext cx, bir:BasicBlock bb, Environment env, s:MemberAccessLExpr lValue, s:Expr rexpr, s:BinaryArithmeticOp|s:BinaryBitwiseOp  op, err:Position pos) returns CodeGenError|StmtEffect{
+    //TODO Move to functions
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: You need to factor out the common code between here and codeGenExpr.

**Review by jclark — CHANGES_REQUESTED** [review]

**Comment by ushirask** [comment]
>      map on LHS (e test in 07-map)
@jclark is compound assignment not allowed with maps?

**Comment by jclark** [comment]
> @jclark is compound assignment not allowed with maps?

It's allowed, but if m has type `map<T>` then `m[k]` will have type `T?` and none of the compound operators allow an operand of `T?`. This will change with https://github.com/ballerina-platform/ballerina-spec/issues/191, but we are not implementing that yet.

File: compiler/modules/front/codeGen.bal [hunk 2]
```diff
@@ -855,6 +843,96 @@ function codeGenAssignToMember(CodeGenContext cx, bir:BasicBlock startBlock, Env
...
+    }
+    else {
+        expr = { bitwiseOp: <s:BinaryBitwiseOp> op, left: lValue, right: rexpr };
+    }
+    return codeGenAssignToVar(cx, startBlock, env, lValue.varName, expr);
+}
+
+function codeGenCompoundAssignToMember(CodeGenContext cx, bir:BasicBlock bb, Environment env, s:MemberAccessLExpr lValue, s:Expr rexpr, s:BinaryArithmeticOp|s:BinaryBitwiseOp op, err:Position pos) returns CodeGenError|StmtEffect {
+    var { result: r0, block: block1 } = check codeGenExpr(cx, bb, env, check cx.foldExpr(env, lValue.container, ()));
+    bir:Register r3;
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: Use meaningful names, consistent with other names in the file.

File: compiler/modules/front/codeGen.bal [hunk 3]
```diff
@@ -855,6 +843,96 @@ function codeGenAssignToMember(CodeGenContext cx, bir:BasicBlock startBlock, Env
...
+    bir:Operand r1;
+    bir:Operand r4;
+    bir:BasicBlock block;
+    if r0 is bir:Register {
+        bir:BasicBlock block3;
+        if t:isSubtypeSimple(r0.semType, t:LIST) {
+            { result: r1, block: block3 } = check codeGenExprForInt(cx, block1, env, check cx.foldExpr(env, lValue.index, t:INT));
+            r3 = cx.createRegister(<t:UniformTypeBitSet>t:simpleArrayMemberType(cx.mod.env, r0.semType));
+            bir:ListGetInsn insn = { result: r3, list: r0, operand: <bir:IntOperand> r1, position: pos };
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: We shouldn't have all these casts of r1.

File: compiler/modules/front/codeGen.bal [hunk 4]
```diff
@@ -855,6 +843,96 @@ function codeGenAssignToMember(CodeGenContext cx, bir:BasicBlock startBlock, Env
...
+            }
+            else {
+                {result: r4, block} = check codeGenBitwiseBinaryExpr(cx, block2, r3, <bir:IntOperand> r2, op);  
+            }
+            bir:ListSetInsn insn1 = { list: <bir:Register> r0, index: <bir:IntOperand> r1, operand: r4, position: lValue.pos };
+            block.insns.push(insn1);
+            return { block };
+        }
+        else {
+            return cx.semanticErr("can only apply member compound assignment to list", pos=pos);
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: Deal with this error case at the beginning of the function without indenting everything below.

File: compiler/modules/front/codeGen.bal [hunk 5]
```diff
@@ -855,6 +843,96 @@ function codeGenAssignToMember(CodeGenContext cx, bir:BasicBlock startBlock, Env
...
+    bir:Operand r4;
+    bir:BasicBlock block;
+    if r0 is bir:Register {
+        bir:BasicBlock block3;
+        if t:isSubtypeSimple(r0.semType, t:LIST) {
+            { result: r1, block: block3 } = check codeGenExprForInt(cx, block1, env, check cx.foldExpr(env, lValue.index, t:INT));
+            r3 = cx.createRegister(<t:UniformTypeBitSet>t:simpleArrayMemberType(cx.mod.env, r0.semType));
+            bir:ListGetInsn insn = { result: r3, list: r0, operand: <bir:IntOperand> r1, position: pos };
+            block3.insns.push(insn);
+            var { result: r2, block: block2 } = check codeGenExpr(cx, block3, env, check cx.foldExpr(env, rexpr, t:ANY));
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: t:ANY is not right. The right type depends on the type of `v[i]`. You need a test case for this:

```
float[] v = [1.0];
v[0] += 1; // contextually expected type for `1` is `float`
```

File: compiler/modules/front/codeGen.bal [hunk 6]
```diff
@@ -855,6 +843,96 @@ function codeGenAssignToMember(CodeGenContext cx, bir:BasicBlock startBlock, Env
...
+        else {
+            return cx.semanticErr("can only apply member compound assignment to list", pos=pos);
+        }
+    }
+    else {
+        return cx.semanticErr("can only apply member access to list or mapping", pos=pos);
+    }
+}
+            
+function codeGenArithmeticBinaryExpr(CodeGenContext cx, bir:BasicBlock bb, bir:Operand lhs, bir:Operand rhs, bir:ArithmeticBinaryOp op, bir:Position pos) returns CodeGenError|ExprEffect {
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: op argument before lhs, rhs arguments

File: compiler/modules/front/codeGen.bal [hunk 7]
```diff
@@ -855,6 +843,96 @@ function codeGenAssignToMember(CodeGenContext cx, bir:BasicBlock startBlock, Env
...
+        bir:StringConcatInsn insn = { operands: pair[1], result };
+        bb.insns.push(insn);
+    }
+    else {
+        return cx.semanticErr(`${op} not supported for operand types`);
+    } 
+    return { result, block: bb };
+}
+
+function codeGenBitwiseBinaryExpr(CodeGenContext cx, bir:BasicBlock bb, bir:IntOperand l, bir:IntOperand r, s:BinaryBitwiseOp op) returns CodeGenError|ExprEffect{
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: Inconsistent names for l/r.

Comment (jclark) [inline compiler/modules/front/codeGen.bal]: Missing space before brace

**Review by jclark — CHANGES_REQUESTED** [review]

**Pull request merged**